From 415470a1599d6c7edc3609873606faf93b89e602 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Tue, 23 Jun 2020 10:41:34 +0800
Subject: [PATCH] DRV_DPM_SW-318: DPM enable reinsertion queue config by
 default

- remove a macro which disabled reinsertion queue config
- Use dp_qos_global_info_get() API to provide values of
  dequeue port and qid to upper layer.

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath.h                  | 2 --
 drivers/net/datapath/dpm/datapath_proc.c             | 6 +-----
 drivers/net/datapath/dpm/gswip31/datapath_misc.c     | 7 ++-----
 drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c | 8 +++++++-
 include/net/datapath_api_qos.h                       | 2 ++
 5 files changed, 12 insertions(+), 13 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 1544f7f5a0e9..145339914778 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -86,8 +86,6 @@
 #define DP_DEBUG(flags, fmt, arg...)
 #endif				/* end of CONFIG_INTEL_DATAPATH_DBG */
 
-#define REINSERTION_NEWQMAP_ENABLE 0
-
 static inline bool is_invalid_port(unsigned int port_id)
 {
 	if (port_id >= MAX_DP_PORTS) {
diff --git a/drivers/net/datapath/dpm/datapath_proc.c b/drivers/net/datapath/dpm/datapath_proc.c
index 605ad403cb66..4129138df3b3 100644
--- a/drivers/net/datapath/dpm/datapath_proc.c
+++ b/drivers/net/datapath/dpm/datapath_proc.c
@@ -284,9 +284,7 @@ int proc_port_dump(struct seq_file *s, int pos)
 	struct inst_info *info = NULL;
 	struct dma_chan_info *dma;
 	struct cqm_port_info *cqm;
-#if REINSERTION_NEWQMAP_ENABLE
 	struct cqm_port_info *cqm_info;
-#endif
 
 	if (!capable(CAP_SYS_PACCT))
 		return -1;
@@ -458,17 +456,15 @@ int proc_port_dump(struct seq_file *s, int pos)
 		   STATS_GET(port->rx_err_drop),
 		   STATS_GET(port->tx_err_drop));
 EXIT:
-#if REINSERTION_NEWQMAP_ENABLE
 	if (pos == 0 && reinsert_deq_port) {
 		cqm_info = get_dp_deqport_info(tmp_inst, reinsert_deq_port);
 		seq_printf(s, "    insertion_qid:     %d\n",
-			   cqm_info->qos_port);
+			   cqm_info->qid[0]);
 		seq_printf(s, "    insertion_deqport: %d\n", reinsert_deq_port);
 		dp_dma_parse_id(cqm_info->dma_chan, &cid, &pid, &nid);
 		seq_printf(s, "    insertion_tx_dma_ch:   0x%x(dma-ctrl=%d,port=%d,channel=%d)\n",
 			   cqm_info->dma_chan, cid, pid, nid);
 	}
-#endif
 
 	if (!seq_has_overflowed(s))
 		pos++;
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 1de2f24db9cc..2970d39752fb 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -1009,7 +1009,6 @@ static void dp_platform_color_table_set(int inst)
 	dp_gswip_remark_dscp_set(GSW_MARKING_DSCP_AF, inst);
 }
 
-#if REINSERTION_NEWQMAP_ENABLE
 static int dp_reinsert_q_map(int inst, struct pmac_port_info *port_info,
 			     struct cbm_cpu_port_data *cpu_data)
 {
@@ -1062,7 +1061,8 @@ static int dp_reinsert_q_map(int inst, struct pmac_port_info *port_info,
 	}
 
 	c_info->dma_chan = re_insertion->dma_chan;
-	c_info->qos_port = q_port.qid;
+	c_info->qos_port = q_port.port_node;
+	c_info->first_qid = q_port.qid;
 	c_info->q_node = q_port.q_node;
 	reinsert_deq_port = q_port.cqe_deq;
 
@@ -1084,7 +1084,6 @@ static int dp_reinsert_q_map(int inst, struct pmac_port_info *port_info,
 
 	return DP_SUCCESS;
 }
-#endif
 
 int dp_platform_queue_set(int inst, u32 flag)
 {
@@ -1239,13 +1238,11 @@ int dp_platform_queue_set(int inst, u32 flag)
 		}
 	}
 
-#if REINSERTION_NEWQMAP_ENABLE
 	/* Re-Insertion use a New Qid -> Deq Port 23 -> Dma1Tx Ch 16 */
 	if (dp_reinsert_q_map(inst, port_info, &cpu_data)) {
 		pr_err("DPM Fail to enable re-insertion Q\n");
 		return -1;
 	}
-#endif
 	return 0;
 }
 
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 13cb4de9d3a6..8c0169c838b9 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -4269,6 +4269,7 @@ int dp_node_reserve(int inst, int ep, struct dp_dev_data *data, int flags)
  */
 int dp_qos_global_info_get_31(struct dp_qos_cfg_info *info, int flag)
 {
+	struct cqm_port_info *cqm_info;
 	struct hal_priv *priv;
 	u32 quanta = 0;
 
@@ -4286,7 +4287,12 @@ int dp_qos_global_info_get_31(struct dp_qos_cfg_info *info, int flag)
 		return DP_FAILURE;
 	}
 	info->quanta = quanta;
-	DP_DEBUG(DP_DBG_FLAG_QOS, "quanta=%d\n", quanta);
+	cqm_info = get_dp_deqport_info(info->inst, reinsert_deq_port);
+	info->reinsert_deq_port = reinsert_deq_port;
+	info->reinsert_qid = cqm_info->qid[0];
+	DP_DEBUG(DP_DBG_FLAG_QOS,
+		   "quanta=%d, reinsert_deq_port = %d, reinsert_qid = %d\n",
+		   quanta, info->reinsert_deq_port, info->reinsert_qid);
 
 	return DP_SUCCESS;
 }
diff --git a/include/net/datapath_api_qos.h b/include/net/datapath_api_qos.h
index 5ee8ca305add..fcc763765791 100644
--- a/include/net/datapath_api_qos.h
+++ b/include/net/datapath_api_qos.h
@@ -1259,6 +1259,8 @@ int dp_ingress_ctp_tc_map_set(struct dp_tc_cfg *tc, int flag);
 struct dp_qos_cfg_info {
 	int inst; /*!< input: dp instance. For SOC side, it is always zero */
 	u32 quanta; /*!< QoS quanta for scheduler */
+	u32 reinsert_deq_port; /*!< dequeue port allocated for reinsertion flow */
+	u32 reinsert_qid; /*!< queue ID allocated for reinsertion flow */
 };
 
 /*!< API dp_qos_global_info_get: Helps to retrieve global QoS
